flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/DataFileRewriteCommitter.java

```diff
@@ -0,0 +1,320 @@
+            .getListState(
+                new ListStateDescriptor<>(
+                    "data-file-updater-in-progress", TypeInformation.of(RewriteFileGroup.class)));
+
+    this.startingSnapshotId = null;
+    Iterable<Long> snapshotIdIterable = startingSnapshotIdState.get();
+    if (snapshotIdIterable != null) {
+      Iterator<Long> snapshotIdIterator = snapshotIdIterable.iterator();
+      if (snapshotIdIterator.hasNext()) {
+        this.startingSnapshotId = snapshotIdIterator.next();
+      }
+    }
+
+    // Has to be concurrent since it is accessed by the CommitService from another thread
+    this.inProgress = Sets.newConcurrentHashSet();
+    Iterable<RewriteFileGroup> inProgressIterable = inProgressState.get();
+    if (inProgressIterable != null) {
+      for (RewriteFileGroup group : inProgressIterable) {
+        inProgress.add(group);
+      }
+    }
+
+    commitInProgress(System.currentTimeMillis());
+
+    this.processed = 0;
+  }
+
+  @Override
+  public void snapshotState(StateSnapshotContext context) throws Exception {
+    super.snapshotState(context);
+
+    startingSnapshotIdState.clear();
+    if (startingSnapshotId != null) {
+      startingSnapshotIdState.add(startingSnapshotId);
+    }
+
+    inProgressState.clear();
+    for (RewriteFileGroup group : inProgress) {
+      inProgressState.add(group);
+    }
+  }
+
+  @Override
+  public void processElement(StreamRecord<DataFileRewriteExecutor.ExecutedGroup> streamRecord) {
+    DataFileRewriteExecutor.ExecutedGroup executedGroup = streamRecord.getValue();
+    try {
+      if (commitService == null) {
+        this.startingSnapshotId = executedGroup.snapshotId();
+        this.commitService =
+            createCommitService(streamRecord.getTimestamp(), executedGroup.groupsPerCommit());
+      }
+
+      commitService.offer(executedGroup.group());
+      inProgress.add(executedGroup.group());
+      ++processed;
+    } catch (Exception e) {
+      LOG.info(
+          LogUtil.MESSAGE_PREFIX + "Exception processing {}",
+          tableName,
+          taskName,
+          taskIndex,
+          streamRecord.getTimestamp(),
+          executedGroup,
+          e);
+      output.collect(TaskResultAggregator.ERROR_STREAM, new StreamRecord<>(e));
+      errorCounter.inc();
+    }
+  }
+
+  @Override
+  public void processWatermark(Watermark mark) throws Exception {
+    try {
+      if (commitService != null) {
+        commitService.close();
+        if (processed != commitService.results().size()) {
+          throw new RuntimeException(
+              String.format(
+                  Locale.ROOT,
+                  LogUtil.MESSAGE_FORMAT_PREFIX + "From %d commits only %d were unsuccessful",
+                  tableName,
+                  taskName,
+                  taskIndex,
+                  mark.getTimestamp(),
+                  processed,
+                  commitService.results().size()));
+        }
+      }
+
+      table.refresh();
```

Comment:
is this refresh needed?

Reply:
Removed the refresh, and the snapshotId from the next log message